substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrMethodRepository.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2021, 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.jfr;
+
+import java.util.Collection;
+import java.util.HashMap;
+
+import org.graalvm.nativeimage.Platform;
+import org.graalvm.nativeimage.Platforms;
+
+import com.oracle.svm.core.SubstrateUtil;
+import com.oracle.svm.core.jdk.JfrStackTraceElement;
+import com.oracle.svm.jfr.traceid.JfrTraceId;
+
+public class JfrMethodRepository implements JfrConstantPool {
+
+    private final JfrMethodHashtable jfrMethodHashtable;
+
+    @Platforms(Platform.HOSTED_ONLY.class)
+    public JfrMethodRepository() {
+        this.jfrMethodHashtable = new JfrMethodHashtable();
+    }
+
+    public void teardown() {
+        jfrMethodHashtable.teardown();
+    }
+
+    public long getMethodId(JfrStackTraceElement jfrStackTraceElement) {
+        return jfrMethodHashtable.putIfAbsent(jfrStackTraceElement);
+    }
+
+    @Override
+    public int write(JfrChunkWriter writer) {
+        return writeMethods(writer);
+    }
+
+    private int writeMethods(JfrChunkWriter writer) {
+        if (jfrMethodHashtable.getSize() == 0) {
+            return 0;
+        }
+
+        writer.writeCompressedLong(JfrTypes.Method.getId());
+        writer.writeCompressedInt(jfrMethodHashtable.getSize());
+
+        for (JfrMethod jfrMethod : jfrMethodHashtable.getMethods()) {
+            writeMethod(writer, jfrMethod);
```

Comment:
I think the hash table should be cleared after iteration so that there isn't potentially unused data in following chunks.

Reply:
Nice catch! Done.